content/post/2025-02-01-lazy/index.Rmd

```diff
@@ -0,0 +1,132 @@
+---
+slug: lazy-meanings
+title: "Lazy introduction to laziness in R" 
+authors: 
+- Maëlle Salmon
+date: "2025-02-01" 
+tags: 
+- package development
+- programming
+output: hugodown::hugo_document
+---
+
+```{r setup, include=FALSE}
+
+knitr::opts_chunk$set(fig.path = "", comment = "")
+# knitr hook to make images output use Hugo options
+knitr::knit_hooks$set(
+  plot = function(x, options) {
+    hugoopts <- options$hugoopts
+    paste0(
+      "{{<figure src=",
+      '"', x, '" ',
+      if (!is.null(hugoopts)) {
+        glue::glue_collapse(
+          glue::glue('{names(hugoopts)}="{hugoopts}"'),
+          sep = " "
+        )
+      },
+      ">}}\n"
+    )
+  }
+)
+
+```
+
+```{r, include = FALSE, cache = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>",
+  out.width = "100%",
+  asciicast_theme = "pkgdown"
+)
+```
+
+In the programming world, laziness can often be a good thing: it is both a human quality that can motivate automation efforts, and a programming concept that avoids wasting resources such as memory.
+Now, when reading code or documentation, seeing the word "lazy" can be confusing, because of its polisemy: it carries several meanings.
+In this post, we will enumerate the different possible definitions of "lazy" in R code.
+
+## Lazy as in lazy evaluation
+
+You might know that R provides lazy evaluation: the arguments of a function are only evaluated if they are accessed.
+In short, you can pass anything as an argument value to a function without any problem as long as the function does not use that value.
+The contrary of lazy evaluation is *eager evaluation*.
+
+The [Advanced R book by Hadley Wickham](https://adv-r.hadley.nz/functions.html#lazy-evaluation) features a very clear introduction to lazy evaluation.
+
+Note that the workhorse of lazy evaluation in base R is a thing called a promise that contains an expression (the recipe for getting a value), an environment (the ingredients that are around), a value.
+The latter is only computed when accessed, and cached once computed.
+
+### What about {future}?
+
+Maybe you have heard of the [future package](https://future.futureverse.org/index.html) by Henrik Bengtsson.
+It provides an implementation in R of futures, a programming concept.
+Its homepage state "In programming, a future is an abstraction for a value that may be available at some point in the future. The state of a future can either be unresolved or resolved."
+
+With futures, you create a future, that is associated to a promise, not the same as the aforementioned base R promise, which is a placeholder for a value and then the value itself.
```

Comment:
this is a very heavy sentence, mainly because of the concepts. I think maybe placing the clarifying part-sentence in paretheses might make it better, as it would make it more clear that the last part of the sentence is an explanation of what a 'promise' is in this context (rather than it now possibly being interpreted as what the aforementioned 'promise' was).

Reply:
adding an example and rephrasing